File: src/Javascript/array/54.螺旋矩阵.js
/*
 * @lc app=leetcode.cn id=54 lang=javascript
 *
 * [54] 螺旋矩阵
 */

// @lc code=start
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
var spiralOrder = function(matrix) {
    let l = 0, t = 0, r = matrix[0].length - 1, b = matrix.length - 1;
    const res = [], count = matrix.length * matrix[0].length;
    let num = 1;
    while (num <= count) {
        for (let i = l; i <= r && num <= count; i++) {
            res.push(matrix[t][i]);
            num++;
        }
        t++;
        for (let i = t; i <= b && num <= count; i++) {
            res.push(matrix[i][r]);
            num++;
        }
        r--;
        for (let i = r; i >= l && num <= count; i--) {
            res.push(matrix[b][i]);
            num++;
        }
        b--;
        for (let i = b; i >= t && num <= count; i--) {
            res.push(matrix[i][l]);
            num++;
        }
        l++;
    }
    return res;
};
// @lc code=end

File: src/Java/practice/first-time/54.螺旋矩阵.java
/*
 * @lc app=leetcode.cn id=54 lang=java
 *
 * [54] 螺旋矩阵
 */

// @lc code=start
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> res = new ArrayList<>();
        int top = 0, bottom = matrix.length - 1, left = 0, right = matrix[0].length - 1;
        int count = (matrix.length ) * (matrix[0].length ), num = 1;
        while (count >= num) {
            for (int i = left; i <= right && count >= num; i++) {
                res.add(matrix[top][i]);
                num++;
            }
            top++;
            for (int i = top; i <= bottom && count >= num; i++) {
                res.add(matrix[i][right]);
                num++;
            }
            right--;
            for (int i = right; i >= left && count >= num; i--) {
                res.add(matrix[bottom][i]);
                num++;
            }
            bottom--;
            for (int i = bottom; i >= top && count >= num; i--) {
                res.add(matrix[i][left]);
                num++;
            }
            left++;
        }
        return res;
    }
}
// @lc code=end

